plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java

```diff
@@ -0,0 +1,242 @@
+import static org.apache.phoenix.util.SchemaUtil.getEscapedArgument;
+
+public class PhoenixMergeSink
+        implements ConnectorMergeSink
+{
+    private final String schemaName;
+    private final String tableName;
+    private final boolean hasRowKey;
+    private final int columnCount;
+    private final List<String> mergeRowIdFieldNames;
+
+    private final ConnectorPageSink insertSink;
+    private final ConnectorPageSink updateSink;
+    private final ConnectorPageSink deleteSink;
+
+    public PhoenixMergeSink(PhoenixClient phoenixClient, RemoteQueryModifier remoteQueryModifier, ConnectorSession session, ConnectorMergeTableHandle mergeHandle, ConnectorPageSinkId pageSinkId)
+    {
+        PhoenixMergeTableHandle phoenixMergeTableHandle = (PhoenixMergeTableHandle) mergeHandle;
+        PhoenixOutputTableHandle phoenixOutputTableHandle = phoenixMergeTableHandle.phoenixOutputTableHandle();
+        this.schemaName = phoenixOutputTableHandle.getSchemaName();
+        this.tableName = phoenixOutputTableHandle.getTableName();
+        this.hasRowKey = phoenixOutputTableHandle.rowkeyColumn().isPresent();
+        this.columnCount = phoenixOutputTableHandle.getColumnNames().size();
+
+        this.insertSink = new JdbcPageSink(session, phoenixOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
+        this.updateSink = createUpdateSink(session, phoenixOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
+
+        ImmutableList.Builder<String> mergeRowIdFieldNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> mergeRowIdFieldTypesBuilder = ImmutableList.builder();
+        RowType mergeRowIdColumnType = (RowType) phoenixMergeTableHandle.mergeRowIdColumnHandle().getColumnType();
+        for (RowType.Field field : mergeRowIdColumnType.getFields()) {
+            checkArgument(field.getName().isPresent(), "Merge row id column filed must have name");
+            mergeRowIdFieldNamesBuilder.add(getEscapedArgument(field.getName().get()));
+            mergeRowIdFieldTypesBuilder.add(field.getType());
+        }
+        this.mergeRowIdFieldNames = mergeRowIdFieldNamesBuilder.build();
+        this.deleteSink = createDeleteSink(session, mergeRowIdFieldTypesBuilder.build(), phoenixClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private ConnectorPageSink createUpdateSink(
+            ConnectorSession session,
+            PhoenixOutputTableHandle phoenixOutputTableHandle,
+            PhoenixClient phoenixClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
+        columnNamesBuilder.addAll(phoenixOutputTableHandle.getColumnNames());
+        columnTypesBuilder.addAll(phoenixOutputTableHandle.getColumnTypes());
+        if (hasRowKey) {
```

Comment:
I'm not really familiar with Jdbc connnectors or Phoenix, but I'm a little surprised that `hasRowKey` can be empty.

Reply:
Trino can allow user create table without explicitly specifying the `rowkeys`, and it will add a column named `ROWKEY` if user doesn't  specify it.
The `rowkeys` can have 2 situations:

1. User specified, the `rowkeys` are the one or more of the user defined columns.
2. Added by Trino, the name is `ROWKEY` and the type is `BIGINT`, value is comes from a `sequence` defined when creating the table.

In situation 1, we need extra care when those cases: 

1. Scan the target table we need added the `ROWKEY` column 
2. Update the target table we should update the `ROWKEY`
3. Insert into the target table we should not assign the value for the `ROWKEY`

And the delete situation is the same as situation 2.

Here add the `rowkeys` is the case 2, update.

Reply:
Thanks for the detailed explanation.